File: clients/go/memory/transport_rest.go
package memory

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// restTransport speaks the hosted Neo4j Agent Memory REST API at
// https://memory.neo4jlabs.com/v1.
//
// It maps the bridge-style Call(method, params) calls to REST endpoints with
// snake_case ↔ camelCase translation on the wire. Methods that have no REST
// equivalent return *NotSupportedError.
type restTransport struct {
	endpoint      string
	apiKey        string
	tokenProvider TokenProvider
	headers       map[string]string
	httpClient    *http.Client
}

func newRestTransport(cfg Config, timeout time.Duration) Transport {
	return &restTransport{
		endpoint:      strings.TrimRight(cfg.Endpoint, "/"),
		apiKey:        cfg.APIKey,
		tokenProvider: cfg.TokenProvider,
		headers:       cfg.Headers,
		httpClient:    &http.Client{Timeout: timeout},
	}
}

func (t *restTransport) Connect(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, t.endpoint+"/conversations?limit=1", nil)
	if err != nil {
		return &ConnectionError{MemoryError{Message: "failed to create request", Cause: err}}
	}
	if err := t.applyHeaders(ctx, req, false); err != nil {
		return err
	}
	resp, err := t.httpClient.Do(req)
	if err != nil {
		return &ConnectionError{MemoryError{Message: fmt.Sprintf("connect to %s failed", t.endpoint), Cause: err}}
	}
	defer resp.Body.Close()
	if resp.StatusCode == 401 || resp.StatusCode == 403 {
		return &AuthenticationError{MemoryError{Message: fmt.Sprintf("auth failed: %d", resp.StatusCode)}}
	}
	if resp.StatusCode >= 500 {
		return &ConnectionError{MemoryError{Message: fmt.Sprintf("server error: %d", resp.StatusCode)}}
	}
	return nil
}

func (t *restTransport) Close(_ context.Context) error { return nil }

type restRoute struct {
	method      string
	path        string
	pathParams  []string
	queryParams []string
	hasBody     bool
	shape       func(raw interface{}, camelParams map[string]interface{}) interface{}
}

var (
	noopRoute        = &restRoute{}
	unsupportedRoute = &restRoute{method: "UNSUPPORTED"}
)

var restRoutes = map[string]*restRoute{
	"setup":          noopRoute,
	"teardown":       noopRoute,
	"clear_all_data": noopRoute,

	"add_message": {method: http.MethodPost, path: "/conversations/{sessionId}/messages", pathParams: []string{"sessionId"}, hasBody: true},
	"get_conversation": {method: http.MethodGet, path: "/conversations/{sessionId}/messages", pathParams: []string{"sessionId"}, queryParams: []string{"limit"},
		shape: func(raw interface{}, p map[string]interface{}) interface{} {
			messages := raw
			if m, ok := raw.(map[string]interface{}); ok {
				if msgs, ok := m["messages"]; ok {
					messages = msgs
				}
			}
			return map[string]interface{}{
				"id":         p["sessionId"],
				"session_id": p["sessionId"],
				"messages":   messages,
			}
		}},
	"list_sessions": {method: http.MethodGet, path: "/conversations", queryParams: []string{"limit"},
		shape: func(raw interface{}, _ map[string]interface{}) interface{} {
			m, ok := raw.(map[string]interface{})
			if !ok {
				return []interface{}{}
			}
			conv, _ := m["conversations"].([]interface{})
			out := make([]interface{}, 0, len(conv))
			for _, c := range conv {
				cm, _ := c.(map[string]interface{})
				if cm == nil {
					continue
				}
				out = append(out, map[string]interface{}{
					"session_id":    cm["id"],
					"message_count": cm["messageCount"],
					"created_at":    cm["createdAt"],
					"updated_at":    cm["updatedAt"],
				})
			}
			return out
		}},
	"search_messages": {method: http.MethodPost, path: "/conversations/{sessionId}/search", pathParams: []string{"sessionId"}, hasBody: true,
		shape: func(raw interface{}, _ map[string]interface{}) interface{} {
			if m, ok := raw.(map[string]interface{}); ok {
				if msgs, ok := m["messages"]; ok {
					return msgs
				}
			}
			return raw
		}},
	"clear_session":            {method: http.MethodDelete, path: "/conversations/{sessionId}", pathParams: []string{"sessionId"}},
	"delete_message":           unsupportedRoute,
	"add_entity":               {method: http.MethodPost, path: "/entities", hasBody: true},
	"search_entities":          {method: http.MethodPost, path: "/entities/search", hasBody: true, shape: extractEntities},
	"add_preference":           unsupportedRoute,
	"add_fact":                 unsupportedRoute,
	"search_preferences":       unsupportedRoute,
	"get_entity_by_name":       unsupportedRoute,
	"get_related_entities":     unsupportedRoute,
	"add_relationship":         unsupportedRoute,
	"merge_duplicate_entities": unsupportedRoute,
	"start_trace":              unsupportedRoute,
	"add_step":                 unsupportedRoute,
	"record_tool_call":         {method: http.MethodPost, path: "/reasoning/tool-calls", hasBody: true},
	"complete_trace":           unsupportedRoute,
	"get_trace_with_steps":     unsupportedRoute,
	"list_traces":              unsupportedRoute,
	"get_tool_stats":           unsupportedRoute,
	"get_similar_traces":       unsupportedRoute,

	// ---- Volume 5 / hosted-native routes -------------------------------
	"create_conversation":       {method: http.MethodPost, path: "/conversations", hasBody: true},
	"list_conversations":        {method: http.MethodGet, path: "/conversations", queryParams: []string{"limit"}, shape: extractConversations},
	"get_conversation_metadata": {method: http.MethodGet, path: "/conversations/{conversationId}", pathParams: []string{"conversationId"}},
	"delete_conversation":       {method: http.MethodDelete, path: "/conversations/{conversationId}", pathParams: []string{"conversationId"}},
	"get_context":               {method: http.MethodGet, path: "/conversations/{conversationId}/context", pathParams: []string{"conversationId"}},
	"bulk_add_messages":         {method: http.MethodPost, path: "/conversations/{conversationId}/messages/bulk", pathParams: []string{"conversationId"}, hasBody: true, shape: extractMessages},
	"get_observations":          {method: http.MethodGet, path: "/conversations/{conversationId}/observations", pathParams: []string{"conversationId"}, queryParams: []string{"limit"}, shape: extractObservations},
	"get_reflections":           {method: http.MethodGet, path: "/conversations/{conversationId}/reflections", pathParams: []string{"conversationId"}, shape: extractReflections},
	"list_entities":             {method: http.MethodGet, path: "/entities", queryParams: []string{"type", "limit"}, shape: extractEntities},
	"get_entity":                {method: http.MethodGet, path: "/entities/{entityId}", pathParams: []string{"entityId"}},
	"update_entity":             {method: http.MethodPut, path: "/entities/{entityId}", pathParams: []string{"entityId"}, hasBody: true},
	"delete_entity":             {method: http.MethodDelete, path: "/entities/{entityId}", pathParams: []string{"entityId"}},
	"set_entity_feedback":       {method: http.MethodPut, path: "/entities/{entityId}/feedback", pathParams: []string{"entityId"}, hasBody: true},
	"get_entity_history":        {method: http.MethodGet, path: "/entities/{entityId}/history", pathParams: []string{"entityId"}},
	"merge_entities":            {method: http.MethodPost, path: "/entities/{sourceId}/merge", pathParams: []string{"sourceId"}, hasBody: true},
	"get_entity_graph":          {method: http.MethodGet, path: "/entities/graph"},
	"explain_step":              {method: http.MethodGet, path: "/reasoning/explain/{stepId}", pathParams: []string{"stepId"}},
	"get_trace_by_conversation": {method: http.MethodGet, path: "/reasoning/trace/{conversationId}", pathParams: []string{"conversationId"}},
	"get_entity_provenance":     {method: http.MethodGet, path: "/reasoning/provenance/{entityId}", pathParams: []string{"entityId"}},
	"record_step":               {method: http.MethodPost, path: "/reasoning/steps", hasBody: true},
	"list_steps":                {method: http.MethodGet, path: "/reasoning/steps", queryParams: []string{"conversationId"}},
	"cypher_query":              {method: http.MethodPost, path: "/query", hasBody: true},
	"list_api_keys":             {method: http.MethodGet, path: "/auth/api-keys", queryParams: []string{"workspaceId"}},
	"create_api_key":            {method: http.MethodPost, path: "/auth/api-keys", hasBody: true},
	"revoke_api_key":            {method: http.MethodDelete, path: "/auth/api-keys/{keyId}", pathParams: []string{"keyId"}},
	"reveal_api_key":            {method: http.MethodGet, path: "/auth/api-keys/{keyId}/reveal", pathParams: []string{"keyId"}, queryParams: []string{"workspaceId"}},
	"refresh_access_token":      {method: http.MethodPost, path: "/auth/refresh", hasBody: true},
}

func extractEntities(raw interface{}, _ map[string]interface{}) interface{} {
	if m, ok := raw.(map[string]interface{}); ok {
		if e, ok := m["entities"]; ok {
			return e
		}
	}
	return raw
}

func extractConversations(raw interface{}, _ map[string]interface{}) interface{} {
	if m, ok := raw.(map[string]interface{}); ok {
		if c, ok := m["conversations"]; ok {
			return c
		}
	}
	return raw
}

func extractMessages(raw interface{}, _ map[string]interface{}) interface{} {
	if m, ok := raw.(map[string]interface{}); ok {
		if x, ok := m["messages"]; ok {
			return x
		}
	}
	return raw
}

func extractObservations(raw interface{}, _ map[string]interface{}) interface{} {
	if m, ok := raw.(map[string]interface{}); ok {
		if x, ok := m["observations"]; ok {
			return x
		}
	}
	return raw
}

func extractReflections(raw interface{}, _ map[string]interface{}) interface{} {
	if m, ok := raw.(map[string]interface{}); ok {
		if x, ok := m["reflections"]; ok {
			return x
		}
	}
	return raw
}

func (t *restTransport) Call(ctx context.Context, method string, params map[string]interface{}, result interface{}) error {
	route, ok := restRoutes[method]
	if !ok {
		return &NotSupportedError{MemoryError{Message: fmt.Sprintf("method %q is not supported by RestTransport", method)}}
	}
	if route == noopRoute {
		return nil
	}
	if route == unsupportedRoute {
		return &NotSupportedError{MemoryError{Message: fmt.Sprintf("method %q has no equivalent in the hosted REST API", method)}}
	}

	camel, _ := snakeToCamel(params).(map[string]interface{})
	if camel == nil {
		camel = map[string]interface{}{}
	}

	consumed := map[string]bool{}

	path := route.path
	for _, name := range route.pathParams {
		v, ok := camel[name]
		if !ok || v == nil || v == "" {
			return &TransportError{
				MemoryError: MemoryError{Message: fmt.Sprintf("missing path param %q for method %q", name, method)},
				StatusCode:  400,
			}
		}
		path = strings.ReplaceAll(path, "{"+name+"}", url.PathEscape(fmt.Sprintf("%v", v)))
		consumed[name] = true
	}

	q := url.Values{}
	for _, name := range route.queryParams {
		v, ok := camel[name]
		if ok && v != nil {
			q.Set(name, fmt.Sprintf("%v", v))
			consumed[name] = true
		}
	}
	queryString := ""
	if len(q) > 0 {
		queryString = "?" + q.Encode()
	}

	var bodyReader io.Reader
	if route.hasBody {
		body := map[string]interface{}{}
		for k, v := range camel {
			if !consumed[k] && v != nil {
				body[k] = v
			}
		}
		buf, err := json.Marshal(body)
		if err != nil {
			return &MemoryError{Message: "failed to marshal body", Cause: err}
		}
		bodyReader = bytes.NewReader(buf)
	}

	fullURL := t.endpoint + path + queryString
	req, err := http.NewRequestWithContext(ctx, route.method, fullURL, bodyReader)
	if err != nil {
		return &ConnectionError{MemoryError{Message: "failed to create request", Cause: err}}
	}
	if err := t.applyHeaders(ctx, req, route.hasBody); err != nil {
		return err
	}

	resp, err := t.httpClient.Do(req)
	if err != nil {
		return &ConnectionError{MemoryError{Message: fmt.Sprintf("request to %s failed", fullURL), Cause: err}}
	}
	defer resp.Body.Close()

	if resp.StatusCode == 401 || resp.StatusCode == 403 {
		return &AuthenticationError{MemoryError{Message: fmt.Sprintf("auth failed: %d", resp.StatusCode)}}
	}
	if resp.StatusCode == 204 {
		return nil
	}

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return &MemoryError{Message: "failed to read response", Cause: err}
	}

	if resp.StatusCode >= 400 {
		var errResp struct {
			Error string `json:"error"`
		}
		if json.Unmarshal(respBody, &errResp) == nil && errResp.Error != "" {
			return &TransportError{
				MemoryError: MemoryError{Message: fmt.Sprintf("%s failed: %s", method, errResp.Error)},
				StatusCode:  resp.StatusCode,
			}
		}
		return &TransportError{
			MemoryError: MemoryError{Message: fmt.Sprintf("%s failed: HTTP %d", method, resp.StatusCode)},
			StatusCode:  resp.StatusCode,
		}
	}

	if result == nil || len(respBody) == 0 {
		return nil
	}

	var raw interface{}
	if err := json.Unmarshal(respBody, &raw); err != nil {
		return &MemoryError{Message: "failed to decode response", Cause: err}
	}

	if route.shape != nil {
		raw = route.shape(raw, camel)
	}
	snake := camelToSnake(raw)

	// Re-marshal/unmarshal to convert into the typed struct the caller passed.
	buf, err := json.Marshal(snake)
	if err != nil {
		return &MemoryError{Message: "failed to re-marshal response", Cause: err}
	}
	if err := json.Unmarshal(buf, result); err != nil {
		return &MemoryError{Message: "failed to decode shaped response", Cause: err}
	}
	return nil
}

func (t *restTransport) applyHeaders(ctx context.Context, req *http.Request, hasBody bool) error {
	if hasBody {
		req.Header.Set("Content-Type", "application/json")
	}
	for k, v := range t.headers {
		req.Header.Set(k, v)
	}
	token := t.apiKey
	if t.tokenProvider != nil {
		fresh, err := t.tokenProvider(ctx)
		if err != nil {
			return &AuthenticationError{MemoryError{Message: "token provider failed", Cause: err}}
		}
		token = fresh
	}
	if token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	return nil
}

File: clients/rlang/neo4j.memory/R/transport_rest.R
#' @title RestTransport — hosted Neo4j Agent Memory REST API
#'
#' @description
#' Speaks the hosted REST API at `https://memory.neo4jlabs.com/v1`. Maps
#' bridge-style `request(method, params)` calls to REST endpoints with
#' snake_case ↔ camelCase translation. Methods that have no REST equivalent
#' raise a "method not supported by RestTransport" error.
#'
#' @export
RestTransport <- R6::R6Class("RestTransport",
  public = list(
    initialize = function(endpoint, api_key = NULL, timeout = 30,
                          token_provider = NULL, headers = NULL) {
      private$endpoint <- sub("/+$", "", endpoint)
      private$api_key <- api_key
      private$timeout <- timeout
      private$token_provider <- token_provider
      private$headers <- if (is.null(headers)) list() else headers
    },

    connect = function() {
      url <- paste0(private$endpoint, "/conversations?limit=1")
      req <- httr2::request(url) |>
        httr2::req_method("GET") |>
        httr2::req_timeout(private$timeout)
      req <- private$apply_auth(req, has_body = FALSE)

      resp <- tryCatch(
        req |> httr2::req_error(is_error = function(r) FALSE) |> httr2::req_perform(),
        error = function(e) NULL
      )
      if (is.null(resp)) return(invisible(self))
      status <- httr2::resp_status(resp)
      if (status == 401L || status == 403L) {
        stop(sprintf("Authentication failed: %d", status), call. = FALSE)
      }
      invisible(self)
    },

    close = function() invisible(NULL),

    request = function(method, params = list()) {
      params <- if (is.null(params)) list() else params
      params <- params[!vapply(params, is.null, logical(1))]
      route <- private$routes[[method]]
      if (is.null(route)) {
        stop(sprintf("Method '%s' is not supported by RestTransport", method),
             call. = FALSE)
      }
      if (identical(route$kind, "noop")) return(NULL)
      if (identical(route$kind, "unsupported")) {
        stop(sprintf("Method '%s' has no equivalent in the hosted REST API",
                     method), call. = FALSE)
      }

      camel <- snake_to_camel(params)
      consumed <- character(0)

      path <- route$path
      for (name in route$path_params %||% character(0)) {
        v <- camel[[name]]
        if (is.null(v) || (is.character(v) && nchar(v) == 0)) {
          stop(sprintf("Missing path param '%s' for method '%s'", name, method),
               call. = FALSE)
        }
        path <- gsub(paste0("\\{", name, "\\}"), utils::URLencode(as.character(v),
          reserved = TRUE), path)
        consumed <- c(consumed, name)
      }

      # Query params on the hosted REST API are snake_case
      # (conversation_id, workspace_id, etc.), NOT camelCase. Look up from
      # the original snake-cased params first; fall back to the camel form
      # if the caller passed it that way. Mark both as consumed so neither
      # leaks into the body.
      qs_pairs <- character(0)
      for (name in route$query_params %||% character(0)) {
        v <- params[[name]]
        if (is.null(v)) {
          v <- camel[[snake_to_camel_key(name)]]
        }
        if (!is.null(v)) {
          qs_pairs <- c(qs_pairs,
            paste0(name, "=", utils::URLencode(as.character(v), reserved = TRUE)))
          consumed <- c(consumed, name, snake_to_camel_key(name))
        }
      }
      qs <- if (length(qs_pairs)) paste0("?", paste(qs_pairs, collapse = "&")) else ""

      url <- paste0(private$endpoint, path, qs)
      req <- httr2::request(url) |>
        httr2::req_method(route$method) |>
        httr2::req_timeout(private$timeout)

      if (isTRUE(route$has_body)) {
        body <- camel[setdiff(names(camel), consumed)]
        req <- req |>
          httr2::req_headers("Content-Type" = "application/json") |>
          httr2::req_body_json(body, auto_unbox = TRUE)
      }

      req <- private$apply_auth(req, has_body = isTRUE(route$has_body))

      resp <- req |>
        httr2::req_error(is_error = function(r) FALSE) |>
        httr2::req_perform()

      status <- httr2::resp_status(resp)
      if (status == 401L || status == 403L) {
        stop(sprintf("Authentication failed: %d", status), call. = FALSE)
      }
      if (status == 204L) return(invisible(NULL))

      body_str <- tryCatch(httr2::resp_body_string(resp), error = function(e) "")
      if (status >= 400L) {
        stop(sprintf("HTTP %d: %s", status, body_str), call. = FALSE)
      }
      if (nchar(body_str) == 0) return(NULL)

      raw <- httr2::resp_body_json(resp, simplifyVector = FALSE)
      if (!is.null(route$shape)) raw <- route$shape(raw, camel)
      camel_to_snake(raw)
    }
  ),

  private = list(
    endpoint = NULL,
    api_key = NULL,
    token_provider = NULL,
    timeout = 30,
    headers = list(),

    apply_auth = function(req, has_body) {
      for (name in names(private$headers)) {
        req <- httr2::req_headers(req, !!name := private$headers[[name]])
      }
      token <- private$api_key
      if (!is.null(private$token_provider)) {
        token <- private$token_provider()
      }
      if (!is.null(token) && nzchar(token)) {
        req <- httr2::req_headers(req, Authorization = paste("Bearer", token))
      }
      req
    },

    routes = list(
      setup = list(kind = "noop"),
      teardown = list(kind = "noop"),
      clear_all_data = list(kind = "noop"),

      add_message = list(method = "POST", path = "/conversations/{sessionId}/messages",
        path_params = c("sessionId"), has_body = TRUE),
      get_conversation = list(method = "GET",
        path = "/conversations/{sessionId}/messages",
        path_params = c("sessionId"), query_params = c("limit"),
        shape = function(raw, p) {
          messages <- raw
          if (is.list(raw) && !is.null(raw$messages)) messages <- raw$messages
          list(id = p$sessionId, sessionId = p$sessionId, messages = messages)
        }),
      list_sessions = list(method = "GET", path = "/conversations",
        query_params = c("limit"),
        shape = function(raw, p) {
          conv <- if (is.list(raw) && !is.null(raw$conversations)) raw$conversations else list()
          lapply(conv, function(c) list(
            sessionId = c$id,
            messageCount = c$messageCount %||% 0L,
            createdAt = c$createdAt,
            updatedAt = c$updatedAt
          ))
        }),
      search_messages = list(method = "POST",
        path = "/conversations/{sessionId}/search",
        path_params = c("sessionId"), has_body = TRUE,
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$messages)) raw$messages else raw
        }),
      clear_session = list(method = "DELETE",
        path = "/conversations/{sessionId}",
        path_params = c("sessionId")),
      delete_message = list(kind = "unsupported"),

      add_entity = list(method = "POST", path = "/entities", has_body = TRUE),
      search_entities = list(method = "POST", path = "/entities/search",
        has_body = TRUE,
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$entities)) raw$entities else raw
        }),
      add_preference = list(kind = "unsupported"),
      add_fact = list(kind = "unsupported"),
      search_preferences = list(kind = "unsupported"),
      get_entity_by_name = list(kind = "unsupported"),
      get_related_entities = list(kind = "unsupported"),
      add_relationship = list(kind = "unsupported"),
      merge_duplicate_entities = list(kind = "unsupported"),

      start_trace = list(kind = "unsupported"),
      add_step = list(kind = "unsupported"),
      record_tool_call = list(method = "POST", path = "/reasoning/tool-calls",
        has_body = TRUE),
      complete_trace = list(kind = "unsupported"),
      get_trace_with_steps = list(kind = "unsupported"),
      list_traces = list(kind = "unsupported"),
      get_tool_stats = list(kind = "unsupported"),
      get_similar_traces = list(kind = "unsupported"),

      # ---- Volume 5 / hosted-native ---------------------------------------
      create_conversation = list(method = "POST", path = "/conversations",
        has_body = TRUE),
      list_conversations = list(method = "GET", path = "/conversations",
        query_params = c("limit"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$conversations)) raw$conversations else raw
        }),
      get_conversation_metadata = list(method = "GET",
        path = "/conversations/{conversationId}",
        path_params = c("conversationId")),
      delete_conversation = list(method = "DELETE",
        path = "/conversations/{conversationId}",
        path_params = c("conversationId")),
      get_context = list(method = "GET",
        path = "/conversations/{conversationId}/context",
        path_params = c("conversationId")),
      bulk_add_messages = list(method = "POST",
        path = "/conversations/{conversationId}/messages/bulk",
        path_params = c("conversationId"), has_body = TRUE,
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$messages)) raw$messages else raw
        }),
      get_observations = list(method = "GET",
        path = "/conversations/{conversationId}/observations",
        path_params = c("conversationId"), query_params = c("limit"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$observations)) raw$observations else raw
        }),
      get_reflections = list(method = "GET",
        path = "/conversations/{conversationId}/reflections",
        path_params = c("conversationId"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$reflections)) raw$reflections else raw
        }),

      list_entities = list(method = "GET", path = "/entities",
        query_params = c("type", "limit"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$entities)) raw$entities else raw
        }),
      get_entity = list(method = "GET", path = "/entities/{entityId}",
        path_params = c("entityId")),
      update_entity = list(method = "PUT", path = "/entities/{entityId}",
        path_params = c("entityId"), has_body = TRUE),
      delete_entity = list(method = "DELETE", path = "/entities/{entityId}",
        path_params = c("entityId")),
      set_entity_feedback = list(method = "PUT",
        path = "/entities/{entityId}/feedback",
        path_params = c("entityId"), has_body = TRUE),
      get_entity_history = list(method = "GET",
        path = "/entities/{entityId}/history",
        path_params = c("entityId")),
      merge_entities = list(method = "POST",
        path = "/entities/{sourceId}/merge",
        path_params = c("sourceId"), has_body = TRUE),
      get_entity_graph = list(method = "GET", path = "/entities/graph"),

      explain_step = list(method = "GET",
        path = "/reasoning/explain/{stepId}",
        path_params = c("stepId")),
      get_trace_by_conversation = list(method = "GET",
        path = "/reasoning/trace/{conversationId}",
        path_params = c("conversationId")),
      get_entity_provenance = list(method = "GET",
        path = "/reasoning/provenance/{entityId}",
        path_params = c("entityId")),
      record_step = list(method = "POST", path = "/reasoning/steps",
        has_body = TRUE),
      list_steps = list(method = "GET", path = "/reasoning/steps",
        query_params = c("conversation_id"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$steps)) raw$steps else raw
        }),
      cypher_query = list(method = "POST", path = "/query", has_body = TRUE),

      list_api_keys = list(method = "GET", path = "/auth/api-keys",
        query_params = c("workspace_id"),
        shape = function(raw, p) {
          if (is.list(raw) && !is.null(raw$keys)) return(raw$keys)
          if (is.list(raw) && !is.null(raw$api_keys)) return(raw$api_keys)
          raw
        }),
      create_api_key = list(method = "POST", path = "/auth/api-keys",
        has_body = TRUE),
      revoke_api_key = list(method = "DELETE", path = "/auth/api-keys/{keyId}",
        path_params = c("keyId")),
      reveal_api_key = list(method = "GET",
        path = "/auth/api-keys/{keyId}/reveal",
        path_params = c("keyId"), query_params = c("workspace_id")),
      refresh_access_token = list(method = "POST", path = "/auth/refresh",
        has_body = TRUE)
    )
  )
)
